### PR TITLE
Functor refactoring

### LINKED ISSUE #285: data race in C01 traversal combine soa
**Describe the bug**
There appears to be a data race in the c01 traversal, see here:
http://vmbungartz10.informatik.tu-muenchen.de/mardyn/blue/rest/organizations/jenkins/pipelines/AutoPas-Multibranch/branches/PR-278/runs/17/nodes/96/steps/200/log/?start=0

**To Reproduce**
* use autopas/autopas-build-archer docker image
* cmake: `CXXFLAGS=-Wno-pass-failed CC=clang-archer CXX=clang-archer++ cmake -G Ninja -DCMAKE_BUILD_TYPE=Release -DAUTOPAS_USE_VECTORIZATION=OFF ..`
* build: `ninja -j 4`
* export TSAN_OPTIONS="ignore_noninstrumented_modules=1" && ctest --verbose

**Expected behavior**
There is no data race

### PR DESCRIPTION
**Pull request by Garfield96**
# Description

Functors now explicitly declare which values are computed and which values are needed. 
- [x] add specialized methods to `ParticleBase` which identify attributes by using the corresponding attribute name (enum)
  - [x] add getter
  - [x] add setter
- [x] remove `#define`
  - [x] `#define AUTOPAS_FUNCTOR_SOALOADER`
  - [x] `#define AUTOPAS_FUNCTOR_SOAEXTRACTOR`
- [x] move `SoALoader`/`SoAExtractor` to class `Functor`
- [x] ~~C04SoA: only write necessary values~~ (Will be done in another PR)
- [x] C01 (combined SoA buffers): only write necessary values (#285)
- [x] use more C++17 in traversals

## Related Pull Requests

- #274 (base branch)

## Resolved Issues

- [x] closes #285

# How Has This Been Tested?

- [x] Jenkins

### CODE REVIEW

File: src/autopas/containers/linkedCells/traversals/C01Traversal.h [hunk 1]
```diff
@@ -99,6 +99,23 @@ class C01Traversal
...
+   * @param appendCell
+   */
+  template <std::size_t... I>
+  inline constexpr void appendNeeded(ParticleCell &cell, ParticleCell &appendCell, std::index_sequence<I...>) {
+    cell._particleSoABuffer.template append<std::get<I>(PairwiseFunctor::getNeededAttr(std::false_type()))...>(
+        appendCell._particleSoABuffer);
+  }
+
+  /**
+   * Resizes all buffers needed for combined SoA buffers.
```
Comment (FG-TUM) [inline src/autopas/containers/linkedCells/traversals/C01Traversal.h]: resize to what size?

File: src/autopas/pairwiseFunctors/Functor.h [hunk 2]
```diff
@@ -20,6 +24,30 @@ namespace autopas {
 template <class Particle>
 class VerletListHelpers;
 
+namespace internal {
+/**
+ * Dummy class to provide empty arrays.
```
Comment (FG-TUM) [inline src/autopas/pairwiseFunctors/Functor.h]: also describe here why and where this hack is needed

File: src/autopas/pairwiseFunctors/LJFunctor.h [hunk 3]
```diff
@@ -149,13 +150,7 @@ class LJFunctor : public Functor<Particle, ParticleCell, typename Particle::SoAA
    * to use it correctly for the global values.
    * @todo: Remove __attribute__((no_sanitize_thread)) when #285 is fixed.
```
Comment (FG-TUM) [inline src/autopas/pairwiseFunctors/LJFunctor.h]: todo can be removed

File: src/autopas/utils/SoA.h [hunk 4]
```diff
@@ -68,16 +68,29 @@ class SoA {
   }
 
   /**
-   * @brief Writes / updates values of attributes for a specific particle.
-   * Appends the other SoA buffer.
-   * @param other other buffer.
+   * @brief Appends the other SoA buffer.
```
Comment (FG-TUM) [inline src/autopas/utils/SoA.h]:
```suggestion
   * Appends the other SoA buffer to this.
```

File: src/autopas/utils/SoA.h [hunk 5]
```diff
@@ -68,16 +68,29 @@ class SoA {
...
+   * @param other Other buffer.
    */
   void append(SoA<SoAArraysType> &other) {
     if (other.getNumParticles() > 0) {
       append_impl(other.soaStorage, std::make_index_sequence<std::tuple_size<SoAArraysType>::value>{});
     }
   }
 
+  /**
+   * @brief Appends the specified attributes from the other SoA buffer.
```
Comment (FG-TUM) [inline src/autopas/utils/SoA.h]:
```suggestion
   * Appends the specified attributes from the other SoA buffer to this.
```

**Comment by SteffenSeckler** [comment]
lgtm error is a false positive and can be ignored (filed bug report https://discuss.lgtm.com/t/c-false-positive-empty-block-without-comment/2183)

**Pull request merged**